lib/traceable.rb

```diff
@@ -0,0 +1,155 @@
+require 'psych'
+
+# Read YAML files and remember where each value came from.
+#
+# Objects behave like normal in most respects. Objects that implement
+# Traceable can be asked for their position by calling file(), line(),
+# column() and byte().
+#
+# If the global variable $traceable_string_output_position is set,
+# to_yaml outputs position information for each atom, separated from
+# its value by the string " =^.^= ".
+#
+# Example:
+#
+#   yaml = "---
+#   foo:
+#     bar: 1
+#     baz:
+#       - gazonk
+#       - foobar"
+#   handler = PositionHandler.new("<STDIN>")
+#   parser = Psych::Parser.new(handler)
+#   handler.parser = parser
+#   parser.parse(yaml)
+#   visitor = PositionVisitor.new
+#   tree = visitor.accept(handler.root)
+#
+#   tree[0]['foo']['bar']
+#   => "1"
+#   tree[0]['foo']['bar'].file
+#   => "<STDIN>"
+#   tree[0]['foo']['bar'].line
+#   => 3
+#   tree[0]['foo']['bar'].column
+#   => 2
+#
+#   puts tree[0].to_yaml
+#   ---
+#   foo:
+#     bar: 1
+#     baz:
+#     - gazonk
+#     - foobar
+#
+#   $dump_with_position = true
```

Comment:
Actually, I'd prefer it to be a class attribute of TraceableString, not a $global.